Spectator camera preview

davidkelly opened the pull request:
A new QML item to allow us to display a network texture (using a QQuickFramebufferObject and friends).  You point a resource://<whatever> at it, and it should render it in a happy little qml window.  For now, this is only rendering at a (hardcoded) 10 hz but that can be adjusted as we see fit

hifi-gustavo commented:
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/791/HighFidelity-Beta-PR10721-9b5b99c2b9f33476b302daaa23f7fab0f1b85eb1.dmg

hifi-gustavo commented:
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/791/HighFidelity-Beta-PR10721-9b5b99c2b9f33476b302daaa23f7fab0f1b85eb1.exe

howard-stearns commented on interface/resources/qml/hifi/SpectatorCamera.qml:
```diff
@@ -189,23 +190,26 @@ Rectangle {
 
         // Spectator Camera Preview
-        Image {
+        Rectangle {
```
In the prototype, I wanted to separate any positioning issues from ResourceImageItem, and so I wrapped the whole thing in a Rectangle that "shouldn't" be necessary.  So, when you get everything working, it would be interesting to get rid of this extra hierarchy.

davidkelly replied:
sounds good.

howard-stearns commented on libraries/display-plugins/src/display-plugins/OpenGLDisplayPlugin.cpp:
```diff
@@ -821,3 +822,35 @@ void OpenGLDisplayPlugin::updateCompositeFramebuffer() {
+        qDebug() << "initial gl error:" << glGetError();
+        GLuint sourceTexture = glBackend->getTextureID(source);
+        GLuint targetTexture = target->texture();
+        GLuint fbo[2] {0, 0};
```
Maybe define the 2 (here, 837, 851) as const size_t N_BUFFERS or some such?

hifi-gustavo commented:
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/813/HighFidelity-Beta-PR10721-5284b8dc0c4ff50fd51a163033dd42dea1b33899.exe

hifi-gustavo commented:
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/813/HighFidelity-Beta-PR10721-5284b8dc0c4ff50fd51a163033dd42dea1b33899.dmg

hifi-gustavo commented:
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/814/HighFidelity-Beta-PR10721-6dd1e6e069a8d282f4bcf5b7b91a393d5ed4510f.dmg

hifi-gustavo commented:
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/814/HighFidelity-Beta-PR10721-6dd1e6e069a8d282f4bcf5b7b91a393d5ed4510f.exe

hifi-gustavo commented:
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/830/HighFidelity-Beta-PR10721-c7e4649ec20135a568d99a9dfbb24ba0cb817834.dmg

hifi-gustavo commented:
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/830/HighFidelity-Beta-PR10721-c7e4649ec20135a568d99a9dfbb24ba0cb817834.exe

hifi-gustavo commented:
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/865/HighFidelity-Beta-PR10721-b4c75c3f877bed841c53cedaacd1564919857468.exe

hifi-gustavo commented:
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/865/HighFidelity-Beta-PR10721-b4c75c3f877bed841c53cedaacd1564919857468.dmg

jherico commented on interface/src/ui/ResourceImageItem.cpp:
```diff
@@ -0,0 +1,80 @@
+//
+
+#include "Application.h"
+#include "ResourceImageItem.h"
```
`ResourceImageItem.h` should be the first header to verify that the header is self-contained.

hifi-gustavo commented:
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/920/HighFidelity-Beta-PR10721-06b59fe13e9f00734ab259b45b8dae050412dffc.exe

hifi-gustavo commented:
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/920/HighFidelity-Beta-PR10721-06b59fe13e9f00734ab259b45b8dae050412dffc.dmg

jherico commented on interface/src/ui/ResourceImageItem.cpp:
```diff
@@ -0,0 +1,107 @@
+    qDebug() << "initial error" << glGetError();
+    glClearColor(0.0f, 0.0f, 0.0f, 0.0f);
+    glClear(GL_COLOR_BUFFER_BIT);
+    QOpenGLFunctions* f = QOpenGLContext::currentContext()->functions();
```
use `auto f = context->versionFunctions<QOpenGLFunctions_4_1_Core>();` and you won't need to get two pointers.

hifi-gustavo commented:
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/1208/HighFidelity-Beta-PR10721-02b712716b34ac4e1bf0e09f0691c256e0ba5cbc.dmg

hifi-gustavo commented:
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/1246/HighFidelity-Beta-PR10721-133af0b522d47372f37d80bff79adeda38945259.dmg

hifi-gustavo commented:
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/1246/HighFidelity-Beta-PR10721-133af0b522d47372f37d80bff79adeda38945259.exe

hifi-gustavo commented:
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/1249/HighFidelity-Beta-PR10721-09d959bdeb0d46726ed349c6936ef40420cff356.dmg

hifi-gustavo commented:
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/1249/HighFidelity-Beta-PR10721-09d959bdeb0d46726ed349c6936ef40420cff356.exe

howard-stearns commented:
so, todo are:
- Austin's comment.
- Aspect ratio/resize (e.g., using the draw black technique)
And we'll treat the needs-stylus problem separately.

hifi-gustavo commented:
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/1256/HighFidelity-Beta-PR10721-455233dc161c88613a8acce96aae785be559dc44.dmg

davidkelly commented:
test this please

hifi-gustavo commented:
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/1259/HighFidelity-Beta-PR10721-455233dc161c88613a8acce96aae785be559dc44.dmg

hifi-gustavo commented:
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/1261/HighFidelity-Beta-PR10721-073eae39b2c4f2596be010471deb021be1ca89a9.dmg

hifi-gustavo commented:
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/1299/HighFidelity-Beta-PR10721-373bd0cc8a07d1b3e0afc036622fa84ed2034e4a.dmg

hifi-gustavo commented:
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/1299/HighFidelity-Beta-PR10721-373bd0cc8a07d1b3e0afc036622fa84ed2034e4a.exe

hifi-gustavo commented:
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/1303/HighFidelity-Beta-PR10721-99bf100ff82e6a3f190b353c5f1425c1e52abe39.dmg

davidkelly commented:
addressed jitter and rendering issues, along with handling the aspect ratio changes.  Also moved Application.h to be included in RenderImageItem.h rather than the cpp, which magically makes all the winsock redefinitions go away.  There are other ways to to that (where we include winsock2.h in particular) but this does the trick and satisfies the request in the CR above.

zfox23 commented on interface/resources/qml/hifi/SpectatorCamera.qml:
```diff
@@ -212,31 +213,36 @@ Rectangle {
             id: spectatorCameraPreview;
+            url: "resource://spectatorCameraFrame";
+            ready: cameraToggleCheckBox.checked;
+            mirrorVertically: true;
```
Any idea why this is necessary?

davidkelly replied:
when you go from visible to not visible and back, it seems to not have ready set properly.  This guaranteed it.

zfox23 commented on interface/src/ui/ResourceImageItem.cpp:
```diff
@@ -0,0 +1,111 @@
+
+    resourceImageItem->setFlag(QQuickItem::ItemHasContents);
+    bool urlChanged = false;
+    if( _url != resourceImageItem->getUrl()) {
```
Strange spacing on that paren

davidkelly replied:
yup I'll fix it

zfox23 commented on interface/src/ui/ResourceImageItem.cpp:
```diff
@@ -0,0 +1,111 @@
+    }
+
+    if (_ready && !_updateTimer.isActive()) {
+        _updateTimer.start(100);
```
Magic number - maybe define this as an interval (ie 1000/10) since we're really defining the framerate of the preview.

zfox23 replied:
Also - any idea how much of a performance hit we take as this variable changes?

davidkelly replied:
true -- crappy hardcoded value...  I'll make a const up at the top for it, with a descriptive name.  I will maybe characterize the perf hit and tweak this in a separate PR.

zfox23 commented on libraries/display-plugins/src/display-plugins/OpenGLDisplayPlugin.cpp:
```diff
@@ -821,3 +822,51 @@ void OpenGLDisplayPlugin::updateCompositeFramebuffer() {
+
+
+        // maintain aspect ratio, filling the width first if possible.  If that makes the height too
+        // much, fill height instead. TODO: only do this when texture changes
```
Very nice. I agree with this scaling method.

zfox23 commented on interface/resources/qml/hifi/SpectatorCamera.qml:
```diff
@@ -212,31 +213,36 @@ Rectangle {
+            url: "resource://spectatorCameraFrame";
+            ready: cameraToggleCheckBox.checked;
+            mirrorVertically: true;
             height: 250;
```
Not a big deal - but because the height is static, isn't it possible we end up letterboxing the output texture unnecessarily? Is that better than changing the height of this element based on the height of the texture?

davidkelly replied:
This is a great question.  In the code, I take the full width and adjust the height, unless that results in more height than I have.  This seemed to be a reasonable approach -- also when I changed the height dynamically it felt 'funny'.  Worth revisiting.

hifi-gustavo commented:
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/1304/HighFidelity-Beta-PR10721-2a3b4b3bfb206816cf278a71683aea277d3dd91e.exe

hifi-gustavo commented:
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/1304/HighFidelity-Beta-PR10721-2a3b4b3bfb206816cf278a71683aea277d3dd91e.dmg

hifi-gustavo commented:
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/1305/HighFidelity-Beta-PR10721-36467983d8ac3cce3cce24a76da6592bda79fa94.dmg

hifi-gustavo commented:
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/1305/HighFidelity-Beta-PR10721-36467983d8ac3cce3cce24a76da6592bda79fa94.exe
